### PR TITLE
[APP-1049]: (feat): Backups V2

### PR DESCRIPTION
**Pull request by walmat**
Fixes APP-1049

### CODE REVIEW

File: src/assets/BackupWarning.png
Comment (benisgold) [inline src/assets/BackupWarning.png:1]: noticed that some images don't have the @2x @3x versions

File: src/components/buttons/rainbow-button/RainbowButtonBackground.tsx [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+import { View } from 'react-native';
+import RadialGradient from 'react-native-radial-gradient';
+import { darkModeThemeColors } from '../../../styles/colors';
+import { useTheme } from '../../../theme/ThemeContext';
+import RainbowButtonTypes from './RainbowButtonTypes';
+import styled from '@/styled-thing';
+import { margin } from '@/styles';
+import { magicMemo } from '@/utils';
+
+const RainbowGradientColorsFactory = (darkMode: boolean, disabled: boolean) => ({
```
Comment (dereknelson) [inline src/components/buttons/rainbow-button/RainbowButtonBackground.tsx:13]: boss had you working overtime at the RainbowGradientColorsFactory

**Review by ibrahimtaveras00 — APPROVED** [review]
All issues have been resolved, QA Passed ✅

File: src/hooks/useManageCloudBackups.ts [hunk 2]
```diff
@@ -1,116 +1,104 @@
...
-  const { navigate } = useNavigation();
+  const [accountDetails, setAccountDetails] = useState<GoogleDriveUserData | undefined>(undefined);
+
+  useEffect(() => {
+    getGoogleAccountUserData()
+      .then(accountDetails => {
+        setAccountDetails(accountDetails ?? undefined);
+      })
+      .catch(error => {
+        logger.error(new RainbowError(`Fetching google account data to display in Backups Section failed`), {
```
Comment (brunobar79) [inline src/hooks/useManageCloudBackups.ts:23]:
```suggestion
        logger.error(new RainbowError(`Error Fetching google account data for Backups Section`), {
```

File: src/hooks/useManageCloudBackups.ts [hunk 3]
```diff
@@ -1,116 +1,104 @@
...
-        cancelButtonIndex: 2,
-        destructiveButtonIndex: 1,
+        cancelButtonIndex: IS_ANDROID ? 2 : 1,
+        destructiveButtonIndex: IS_ANDROID ? 0 : 1,
         options: buttons,
-        title: `Manage ${cloudPlatform} Backups`,
+        title: getTitleForPlatform(),
       },
-      async (buttonIndex: number) => {
-        if (buttonIndex === 0) {
```
Comment (brunobar79) [inline src/hooks/useManageCloudBackups.ts:27]: are we removing the option to restore and older backup?

File: src/keychain/index.ts [hunk 4]
```diff
@@ -97,14 +96,23 @@ export async function get(key: string, options: KeychainOptions = {}): Promise<R
             logger.debug(
               `keychain: decrypting private data on Android`,
               {
-                key,
+                extra: {
+                  key,
+                  password: result.password,
+                },
```
Comment (brunobar79) [inline src/keychain/index.ts:102]: Please do not log secrets

File: src/keychain/index.ts [hunk 5]
```diff
@@ -97,14 +96,23 @@ export async function get(key: string, options: KeychainOptions = {}): Promise<R
...
               },
               logger.DebugContext.keychain
             );
 
             const pin = options.androidEncryptionPin || (await authenticateWithPIN());
+            logger.log('keychain: using pin to decrypt cipher', { pin });
```
Comment (brunobar79) [inline src/keychain/index.ts:108]: same

File: src/keychain/index.ts [hunk 6]
```diff
@@ -97,14 +96,23 @@ export async function get(key: string, options: KeychainOptions = {}): Promise<R
...
             const decryptedValue = await encryptor.decrypt(pin, result.password);
 
+            logger.log('keychain: decrypted value', {
```
Comment (brunobar79) [inline src/keychain/index.ts:111]: same

File: src/model/backup.ts [hunk 7]
```diff
@@ -60,6 +89,131 @@ async function extractSecretsForWallet(wallet: RainbowWallet) {
...
+  try {
+    /**
+     * Loop over all keys and decrypt if necessary for android
+     * if no latest backup, create first backup with all secrets
+     * if latest backup, update updatedAt and add new secrets to the backup
+     */
+
+    const allKeys = await keychain.loadAllKeys();
+    if (!allKeys) {
+      onError?.('No keys found. Please try again.'); // TODO: i18n this
```
Comment (brunobar79) [inline src/model/backup.ts:127]: Still TODO?

File: src/model/backup.ts [hunk 8]
```diff
@@ -60,6 +89,131 @@ async function extractSecretsForWallet(wallet: RainbowWallet) {
...
+        };
+      });
+
+      await Promise.all(promises);
+      updatedBackupFile = await encryptAndSaveDataToCloud(data, password, `backup_${now}.json`);
+    } else {
+      // if we have a latest backup file, we need to update the updatedAt and add new secrets to the backup file..
+      const backup = await getDataFromCloud(password, latestBackup);
+      if (!backup) {
+        onError?.('Backup not found. Please try again.'); // TODO: i18n this
```
Comment (brunobar79) [inline src/model/backup.ts:173]: same

File: src/model/backup.ts [hunk 9]
```diff
@@ -268,28 +430,85 @@ export async function restoreCloudBackup({
...
+
+      // if the entry isn't a pkey or seed phrase let's continue on...
+      if (!theKeyIsASeedPhrase && !theKeyIsAPrivateKey) {
+        continue;
+      }
+
+      const valueStr = backedUpData[key];
+      const parsedValue = JSON.parse(valueStr);
+
+      let secretPhraseOrOldAndroidBackupPrivateKey: string | any; // TODO: Strengthen this type
```
Comment (brunobar79) [inline src/model/backup.ts:468]: We have a type for pkeys

**Review by brunobar79 — APPROVED** [review]
Good stuff! Added minor comments but LGTM!

File: src/handlers/walletReadyEvents.ts [hunk 10]
```diff
@@ -50,31 +55,29 @@ export const runWalletBackupStatusChecks = () => {
...
-  }
+    let stepType: string = WalletBackupStepTypes.no_provider;
+    if (backupProvider === walletBackupTypes.cloud) {
+      stepType = WalletBackupStepTypes.backup_now_to_cloud;
+    } else if (backupProvider === walletBackupTypes.manual) {
+      stepType = WalletBackupStepTypes.backup_now_manually;
+    }
 
-  // otherwise, show the BackupSheet redirecting to the WalletSelectionList
-  IS_TESTING !== 'true' &&
```
Comment (skylarbarrera) [inline src/handlers/walletReadyEvents.ts:68]: do we want to remove this? could mess up e2e flows

**Review by skylarbarrera — APPROVED** [review]
🫡

**Pull request merged**